docs/adasum_user_guide.rst

```diff
@@ -0,0 +1,216 @@
+.. image:: media/7220c70747b40ab58fce2dc246958218.png
+
+Ring
+--------------------------
+
+On specifically configured machines (`DGX1 <https://www.nvidia.com/en-us/data-center/dgx-1/>`_ nodes with 8 GPUs each), the Ring mode can be used instead of the pure CPU mode. This mode is identical to the pure CPU mode for inter-node communication, but is able to do intra-node communication without going through the CPU. It does this by utilizing CUDA-aware MPI (OpenMPI built with `UCX <https://www.openucx.org/>`_ support) in order to allow direct GPU to GPU communication within nodes. This results in identical convergence benefits to pure CPU mode, but much better throughput on nodes that support it.
+
+Ring mode is currently supported only on **DGX1** nodes having 8 GPUs each.
+
+.. image:: media/4920a765a77fa6eeca28c5aceaa405ec.png
+
+Hierarchical
+--------------------------
+
+In cases where the hardware does not support Ring mode, but throughput higher than that of the pure CPU mode is desired, the hierarchical mode can be used instead.
+
+The hierarchical mode functions similar to the Ring mode, except for using NCCL to do regular averaging intra-node, instead of using CUDA-aware MPI to do an AdaSum-like ring. Note that hierarchical also works on any hardware configuration, and is not limited to DGX1s.
+
+In practice, hierarchical yields the best throughput, but lowers the convergence benefits of AdaSum due to some of the ops being regular averaging. As a rule of thumb, typically the convergence benefit degradation is insignificant on clusters with large numbers of nodes (\>=8), as in that case there are enough inter-node AdaSum ops being performed. This is the ideal Hierarchical scenario.
+
+The other reason to use Hierarchical even on smaller clusters is when Ring mode is not supported, and CPU mode throughput is simply too low to be viable. Note that in these cases the convergence benefits compared to not using AdaSum at all might be minor.
+
+The learning rate that should be used is equal to the best learning rate for a single worker (GPU) scaled by the number of GPUs locally on a node. On very large clusters, scaling this even more by another factor of 1.5-2.0x might give better results but is not guaranteed and should be tried only if scaling by just the local size is not sufficient for good convergence
+
+.. image:: media/a254d38d0e56319c0507a16ea09df959.png
+
+Modification to the code
+===========================
+
+A new distributed optimizer has been added to both TensorFlow and Pytorch to support the AdaSum algorithm.
+
+An optional parameter op has been added to DistributedOptimizer and allreduce API for users to specify which operation to perform.
+When **op=hvd.AdaSum** is specified, the new optimizer will be used.
+
+TensorFlow
+--------------------------
+
+-   DistributedOptimizer
+
+.. code-block:: python
+
+    opt = tf.train.AdamOptimizer(0.001)
+    opt = hvd.DistributedOptimizer(opt, backward_passes_per_step=5, op=hvd.AdaSum)
```

Comment:
Can you comment on when it makes sense to set `backward_passes_per_step`?

Reply:
AdaSum is useful for scaling to large batch sizes, and that can be done with a limited GPU memory by using gradient accumulation using `backward_passes_per_step`. I have added the same explanation in the userguide, thanks!